cartesian_product compile error

Hi!

Seems like the latest changes 0d5600e broke the following code:

```
#include <range/v3/all.hpp>

int main() {
    const auto xs = ranges::views::indices( size_t{ 0 }, size_t{ 10 } );
    const auto ys = ranges::views::indices( size_t{ 0 }, size_t{ 10 } );

    for ( auto r : ranges::views::cartesian_product( ys, xs ) )
    {
    }
}
```

tested with gcc 9.1 and clang 8.0
<details>
<summary>Click to see the error</summary>
<p>

```
[voivoid@voivoidpc Sandbox]$ clang++ -I/home/voivoid/Libs/range-v3/include ./main.cpp 
In file included from ./main.cpp:1:
In file included from /home/voivoid/Libs/range-v3/include/range/v3/all.hpp:25:
In file included from /home/voivoid/Libs/range-v3/include/range/v3/view.hpp:27:
/home/voivoid/Libs/range-v3/include/range/v3/view/cartesian_product.hpp:198:24: error: no viable conversion from returned value of type '::concepts::detail::enable_if_t<typename basic_iterator<cursor>::difference_type,
      static_cast<bool>(detail::sized_sentinel_for_cursor<cursor, cursor>) || CPP_false(::concepts::detail::xNil{})>' (aka 'ranges::detail::diffmax_t') to function return type 'std::intmax_t' (aka 'long')
                return std::get<0>(that.its_) - std::get<0>(its_);
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/voivoid/Libs/range-v3/include/range/v3/view/cartesian_product.hpp:203:35: note: in instantiation of member function 'ranges::cartesian_product_view<ranges::iota_view<unsigned long, unsigned long>, ranges::iota_view<unsigned long, unsigned long>
      >::cursor<false>::distance_' requested here
                std::intmax_t d = distance_(that, meta::size_t<N - 1>{});
                                  ^
/home/voivoid/Libs/range-v3/include/range/v3/view/cartesian_product.hpp:343:24: note: in instantiation of function template specialization 'ranges::cartesian_product_view<ranges::iota_view<unsigned long, unsigned long>, ranges::iota_view<unsigned long,
      unsigned long> >::cursor<false>::distance_<2>' requested here
                return distance_(that, meta::size_t<sizeof...(Views)>{});

```

</p>
</details>

--- a/include/range/v3/view/cartesian_product.hpp
+++ b/include/range/v3/view/cartesian_product.hpp
@@ -140,6 +140,8 @@ namespace ranges
             friend cursor<true>;
             template<typename T>
             using constify_if = meta::const_if_c<IsConst_, T>;
+            using difference_type = common_type_t<std::intmax_t, range_difference_t<Views>...>;
+
             constify_if<cartesian_product_view> * view_;
             std::tuple<iterator_t<constify_if<Views>>...> its_;
 
@@ -193,37 +195,36 @@ namespace ranges
                 return std::get<N - 1>(its_) == std::get<N - 1>(that.its_) &&
                        equal_(that, meta::size_t<N - 1>{});
             }
-            std::intmax_t distance_(cursor const & that, meta::size_t<1>) const
+            difference_type distance_(cursor const & that, meta::size_t<1>) const
             {
-                return std::get<0>(that.its_) - std::get<0>(its_);
+                return difference_type{std::get<0>(that.its_) - std::get<0>(its_)};
             }
             template<std::size_t N>
-            std::intmax_t distance_(cursor const & that, meta::size_t<N>) const
+            difference_type distance_(cursor const & that, meta::size_t<N>) const
             {
-                std::intmax_t d = distance_(that, meta::size_t<N - 1>{});
-                d *= ranges::distance(std::get<N - 1>(view_->views_));
-                d += std::get<N - 1>(that.its_) - std::get<N - 1>(its_);
-                return d;
+                difference_type const d = distance_(that, meta::size_t<N - 1>{});
+                auto const scale = ranges::distance(std::get<N - 1>(view_->views_));
+                auto const increment = std::get<N - 1>(that.its_) - std::get<N - 1>(its_);
+                return difference_type{d * scale + increment};
             }
-            template<typename Diff>
-            void advance_(meta::size_t<0>, Diff)
+            void advance_(meta::size_t<0>, difference_type)
             {
                 RANGES_EXPECT(false);
             }
             RANGES_DIAGNOSTIC_PUSH
             RANGES_DIAGNOSTIC_IGNORE_DIVIDE_BY_ZERO
-            template<std::size_t N, typename Diff>
-            void advance_(meta::size_t<N>, Diff n)
+            template<std::size_t N>
+            void advance_(meta::size_t<N>, difference_type n)
             {
                 if(n == 0)
                     return;
 
                 auto & i = std::get<N - 1>(its_);
                 auto const my_size =
-                    static_cast<Diff>(ranges::size(std::get<N - 1>(view_->views_)));
+                    static_cast<difference_type>(ranges::size(std::get<N - 1>(view_->views_)));
                 auto const first = ranges::begin(std::get<N - 1>(view_->views_));
 
-                Diff const idx = static_cast<Diff>(i - first);
+                auto const idx = static_cast<difference_type>(i - first);
                 RANGES_EXPECT(0 <= idx);
                 RANGES_EXPECT(idx < my_size || (N == 1 && idx == my_size && n < 0));
                 RANGES_EXPECT(n < INTMAX_MAX - idx);
@@ -342,10 +343,9 @@ namespace ranges
             {
                 return distance_(that, meta::size_t<sizeof...(Views)>{});
             }
-            template<typename Diff>
-            auto advance(Diff n) -> CPP_ret(void)( //
-                requires cartesian_produce_view_can_random<IsConst, Views...> &&
-                    detail::integer_like_<Diff>)
+            CPP_member
+            auto advance(difference_type n) -> CPP_ret(void)( //
+                requires cartesian_produce_view_can_random<IsConst, Views...>)
             {
                 advance_(meta::size_t<sizeof...(Views)>{}, n);
             }

--- a/test/view/cartesian_product.cpp
+++ b/test/view/cartesian_product.cpp
@@ -23,6 +23,7 @@
 #include <range/v3/view/chunk.hpp>
 #include <range/v3/view/empty.hpp>
 #include <range/v3/view/filter.hpp>
+#include <range/v3/view/indices.hpp>
 #include <range/v3/view/iota.hpp>
 #include <range/v3/view/reverse.hpp>
 #include <range/v3/view/take_exactly.hpp>
@@ -253,6 +254,18 @@ void test_bug_1269()
     }
 }
 
+void test_bug_1279()
+{
+    // https://github.com/ericniebler/range-v3/issues/1279
+    auto const xs = ranges::views::indices(std::size_t{0}, std::size_t{10});
+    auto const ys = ranges::views::indices(std::size_t{0}, std::size_t{10});
+
+    for(auto r : ranges::views::cartesian_product(ys, xs))
+    {
+        (void) r;
+    }
+}
+
 int main()
 {
     int some_ints[] = {0,1,2,3};
@@ -307,6 +320,7 @@ int main()
     test_bug_919();
     test_bug_978();
     test_bug_1269();
+    test_bug_1279();
 
     return test_result();
 }

